src/SignalR/clients/csharp/Client.SourceGenerator/src/CallbackRegistrationGenerator.Emitter.cs

```diff
@@ -0,0 +1,192 @@
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.AspNetCore.SignalR.Client.SourceGenerator
+{
+    internal partial class CallbackRegistrationGenerator
+    {
+        public class Emitter
+        {
+            private readonly SourceProductionContext _context;
+            private readonly SourceGenerationSpec _spec;
+
+            public Emitter(SourceProductionContext context, SourceGenerationSpec spec)
+            {
+                _context = context;
+                _spec = spec;
+            }
+
+            public void Emit()
+            {
+                if (string.IsNullOrEmpty(_spec.RegisterCallbackProviderClassAccessibility) ||
+                    string.IsNullOrEmpty(_spec.RegisterCallbackProviderMethodAccessibility) ||
+                    string.IsNullOrEmpty(_spec.RegisterCallbackProviderClassName) ||
+                    string.IsNullOrEmpty(_spec.RegisterCallbackProviderMethodName) ||
+                    string.IsNullOrEmpty(_spec.RegisterCallbackProviderTypeParameterName) ||
+                    string.IsNullOrEmpty(_spec.RegisterCallbackProviderConnectionParameterName))
+                {
+                    return;
+                }
+
+                // Generate extensions and other user facing mostly-static code in a single source file
+                EmitExtensions();
+                // Generate specific callback registration methods in their own source file for each provider type
+                foreach (var typeSpec in _spec.Types)
+                {
+                    EmitRegistrationMethod(typeSpec);
+                }
+            }
+
+            private void EmitExtensions()
+            {
+                var registerProviderBody = new StringBuilder();
+
+                // Generate body of RegisterCallbackProvider<T>
+                foreach (var typeSpec in _spec.Types)
+                {
+                    var methodName = $"register{typeSpec.FullyQualifiedTypeName.Replace(".", string.Empty)}";
+                    var fqtn = typeSpec.FullyQualifiedTypeName;
+                    registerProviderBody.AppendLine($@"
+            if(typeof({_spec.RegisterCallbackProviderTypeParameterName}) == typeof({fqtn}))
+            {{
+                return (System.IDisposable) new CallbackProviderRegistration({methodName}({_spec.RegisterCallbackProviderConnectionParameterName}, ({fqtn}) provider));
+            }}");
+                }
+
+                // Generate RegisterCallbackProvider<T> extension method and CallbackProviderRegistration class
+                // RegisterCallbackProvider<T> is used by end-user to register their callback provider types
+                // CallbackProviderRegistration is a private implementation of IDisposable which simply holds
+                //  an array of IDisposables acquired from registration of each callback method from HubConnection
+                var extensions = $@"// Generated by Microsoft.AspNetCore.Client.SourceGenerator
```

Comment:
I was actually curious what we should do for license header for these.. Generally I think when you use a tool to produce something, you have ownership over the product. So I'm wondering if the MIT License header is OK or if we need something looser here?

Reply:
> So I'm wondering if the MIT License header is OK or if we need something looser here?

Yeah, I was just grabbing what we do for our files, I don't think I've seen other generators add a license so that's probably fine to keep out.

Reply:
Updated the generated headers with `// <auto-generated>` block and nullable directive.